### PR TITLE
feat(router-replay): support metadata and moe topk

### PR DESCRIPTION
**Pull request by SolenoidWGT**
## Summary

One-line summary: Add datum-level Router Replay metadata contracts with opaque payload references, training metadata passthrough, and MoE top-k index preservation in sampling responses.

Related Issue: N/A

Related design: RFC-0001 Weaver RL Router Replay training loop

## Changes

- Add Router Replay SDK contract types and exports, including `RouterReplayIndices`, `RouterReplayMetadata`, `RouterReplayModelConfig`, schema constants, URI builders, and materialization helpers.
- Move Router Replay to a datum-level contract: callers attach `Datum.metadata["router_replay"]`, while request-level `router_replay` and `metadata["router_replay"]` are rejected with clear `ValueError`s.
- Use opaque replay references for SDK-visible Router Replay metadata via `sample_ref`, `index_set_uri`, and `manifest_uri`; inline SDK-visible indices, `sample_index`, and `index_uri` are rejected in the final metadata payload.
- Add generic `PayloadRef` helpers for large payload references, with local/GPFS/filesystem materialization for JSON and `torch.save` payloads when explicit inspection is required.
- Extend `Datum` serialization to preserve optional per-datum metadata without mixing it into `loss_fn_inputs`.
- Extend `TrainingClient.forward()` and `TrainingClient.forward_backward()` to pass generic top-level request metadata while keeping Router Replay metadata per datum.
- Extend `SamplingClient.sample()` with the top-level `return_moe_topk_indices` request flag and preserve non-null `moe_topk_indices` in normalized sequence results.
- Add focused unit tests for Router Replay types, payload refs, sampling MoE top-k indices, datum metadata serialization, and training metadata request behavior.

## Change Flow

```mermaid
flowchart TD
    A["SamplingClient.sample(return_moe_topk_indices=True)"] --> B["Sampling request body"]
    B --> C["Weaver sampling service"]
    C --> D["Response sequences with optional moe_topk_indices"]
    D --> E["SamplingClient._normalize_sample_result"]
    E --> F["Caller stores or receives Router Replay top-k refs"]

    F --> G["Caller builds Datum"]
    G --> H{"Datum metadata includes router_replay?"}
    H -- "Yes" --> I["Datum.metadata['router_replay'] uses schema v1 and opaque refs"]
    H -- "No" --> J["Regular training datum"]
    I --> K["TrainingClient.forward / forward_backward"]
    J --> K
    K --> L{"Request-level Router Replay metadata?"}
    L -- "Yes" --> M["Raise ValueError and require datum.metadata"]
    L -- "No" --> N["Serialize data plus optional generic request metadata"]
    N --> O["POST forward-passes or forward-backward-passes"]
    O --> P["Trainer consumes regular data or opaque Router Replay refs"]

    I --> Q{"Debug materialization requested?"}
    Q -- "Default" --> R["Return non-materialized ref summary"]
    Q -- "trusted=True or env enabled" --> S["Resolve manifest and shards through PayloadRef helpers"]
    S --> T{"Readable local/GPFS ref?"}
    T -- "Yes" --> U["Return inspectable indices"]
    T -- "No" --> V["Raise materialization error"]
```

Flow notes:

- Sampling requests only include `return_moe_topk_indices` when explicitly requested, and normalized results preserve `moe_topk_indices` only when the service returns a non-null value.
- Training requests keep generic request metadata at the top level, but Router Replay metadata is intentionally scoped to each `Datum`.
- `RouterReplayMetadata` builds `RECORD` and `REPLAY` payloads around opaque refs so normal SDK flows do not materialize large index payloads.
- Debug helpers return ref summaries by default; trusted materialization can resolve manifests and local/GPFS shard payloads, and surfaces load failures as explicit errors.

## Correctness Tests

- [x] Unit tests added or updated
- [x] Relevant unit tests executed

Test commands:

```bash
python -m pytest tests/test_router_replay_types.py tests/test_sampling_moe_topk_indices.py tests/test_training_metadata.py
git diff --check main...HEAD
```

Test results:

- `30 passed in 13.41s`
- `git diff --check main...HEAD` completed with no output.

## Performance Tests

- [x] No performance impact
- [ ] Performance testing completed
- [ ] Performance test document provided

Performance test document:

- N/A

Performance conclusion:

- This PR changes SDK request/response serialization and typed metadata helpers only. It does not change server-side compute, database access, caching, batching, concurrency, or rendering paths. Large Router Replay payloads remain opaque by default and are materialized only through explicit trusted debug helpers.

## Risks And Rollback

Risks:

- Callers using the earlier draft request-level Router Replay API must move Router Replay data to `Datum.metadata["router_replay"]` and use opaque refs instead of inline SDK-visible indices.
- Debug materialization depends on local/GPFS-readable refs and the `WEAVER_PAYLOAD_REF_ROOT` or `WEAVER_ROUTER_REPLAY_REF_ROOT` environment when relative or `weaver://` paths need resolution.
- `return_moe_topk_indices=True` can make sampling responses larger when the service returns router top-k indices, so callers should enable it only when needed for replay workflows.

Rollback:

- Revert this PR to restore the previous SDK payload shape.
- Until the revert is deployed, callers can avoid the new paths by not passing `return_moe_topk_indices` and not attaching `Datum.metadata["router_replay"]`.

### CODE REVIEW

**Review by void-main — APPROVED** [review]
LGTM
